### PR TITLE
58 type annotation on other nodes

### PR DESCRIPTION
**Pull request by yuanzexiong**
## Description

Annotation on other nodes (All things left in the 58)

Closes #58

## Changes

1. Changed compileAstToExpression to include the type param
2. Added files that handles different types of nodes
3. Added switches cases to call these functions and annotate them (and their children)

## Deletes Source Branch?

Yes:    Work is done, the branch can be safely deleted.

## How Many Approvals?

* Approval of all (Controversial changes that change the entire project)

### CODE REVIEW

File: src/expressions/operators/types/InstanceOfOperator.ts [hunk 1]
```diff
@@ -9,8 +10,19 @@ class InstanceOfOperator extends Expression {
...
+		expression: Expression,
+		typeTest: Expression,
+		multiplicity: string,
+		type: SequenceType
+	) {
+		super(
+			expression.specificity,
+			[expression],
+			{ canBeStaticallyEvaluated: false },
+			undefined,
```
Comment (FBroy) [inline src/expressions/operators/types/InstanceOfOperator.ts]: The undefined should be false

File: src/typeInference/annotateAST.ts [hunk 2]
```diff
@@ -146,6 +159,38 @@ export function annotate(ast: IAST, staticContext: StaticContext): SequenceType
...
+			const elseClause = annotate(
+				astHelper.getFirstChild(astHelper.getFirstChild(ast, 'elseClause'), '*'),
+				staticContext
+			);
+			return annotateIfThenElseExpr(ast, thenClause, elseClause);
+		}
+		case 'instanceOfExpr': {
+			annotate(astHelper.getFirstChild(ast, 'argExpr'), staticContext);
+			annotate(astHelper.getFirstChild(ast, 'sequenceType'), staticContext);
+			annotateInstanceOfExpr(ast);
```
Comment (FBroy) [inline src/typeInference/annotateAST.ts]: This should be returned

File: src/typeInference/annotateAST.ts [hunk 3]
```diff
@@ -179,12 +224,79 @@ export function annotate(ast: IAST, staticContext: StaticContext): SequenceType
...
+		case 'typeSwitchExpr':
+			const arg = annotate(astHelper.getFirstChild(ast, 'argExpr') as IAST, staticContext);
+			const clauses = astHelper
+				.getChildren(ast, 'typeswitchExprCaseClause')
+				.map((a) => annotate(a, staticContext));
+			const defaultCase = annotate(
+				astHelper.getFirstChild(ast, 'typeSwitchExprDefaultClause') as IAST,
+				staticContext
+			);
+			annotateTypeSwitchOperator(ast);
```
Comment (FBroy) [inline src/typeInference/annotateAST.ts]: This should be returned

**Review by FBroy — DISMISSED** [review]

File: src/typeInference/annotateArrowExpr.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+	const func = astHelper.getFirstChild(ast, 'EQName');
+
+	// There may be no name for the function
+	if (!func) {
+		return undefined;
+	}
+
+	// Sometimes there is no prefix given, hence we need to check for that case and give an empty prefix
+	let functionName;
```
Comment (HectorPeeters) [inline src/typeInference/annotateArrowExpr.ts]: Maybe typing these would make the code a bit more readable

File: src/typeInference/annotateDynamicFunctionInvocationExpr.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { SequenceType } from '../expressions/dataTypes/Value';
+import StaticContext from '../expressions/StaticContext';
+import { IAST } from '../parsing/astHelper';
+
+/**
+ * At this moment there is no way to infre the return type of this function as
```
Comment (HectorPeeters) [inline src/typeInference/annotateDynamicFunctionInvocationExpr.ts]: Infre!

**Review by HectorPeeters — APPROVED** [review]

File: src/typeInference/annotateDynamicFunctionInvocationExpr.ts (hunk 5 above)
Comment (FBroy) [reply src/typeInference/annotateDynamicFunctionInvocationExpr.ts]: Well yes, we are doing type infrerence here.

Comment (maanpalle) [reply src/typeInference/annotateDynamicFunctionInvocationExpr.ts]: gooey said we did

**Review by FBroy — APPROVED** [review]
Once Vincent is done with the documentation, we're ready to merge! 
Good job everyone!

**Comment by yuanzexiong** [comment]
Reviewing large pull request is really a pain on GitHub web, perhaps VS Code compare branch is a bit better.

**Review by maanpalle — APPROVED** [review]
Good job everyone

**Pull request merged**